### DISCUSSION TITLE
Use of MSI / SPN / AZCLI

### DISCUSSION THREAD

**Discussion by rajsshah86** [Azure/azure-storage-fuse#1396] [Q&A] [answered]
Hi 
I have few questions :
We have Premium storage account of type BlockBlob storage and HNS enabled.
- What should be the Endpoint value in the config file ? should it be blob or dfs ? The type set is adls . In the logs it seems like it uses dfs endpoint even if we mention blob endpoint .
- The MSI login option : does this only applicable to the case where the MSI is assigned to an azure resource like VM ? Because a user defined msi also supports federated credentials . How can we use that to login via msi and use blobfuse ?
- For SPN authentication as well , is there a way to authenticate via OIDC (federated credentials) ? 
- When is the plan to release the azcli login option ? I can see that it has not officially been released yet even though the code is there :) )

**Comment by souravgupta-msft** [comment, marked as answer]
Hi @rajsshah86. Thanks for reaching out. Please find the reply below.

- If the account type is `adls`, then the endpoint should be `dfs`. If adls is set as account type and you haven't specified the endpoint field in the config, it is auto computed to `accountName.dfs.core.windows.net`. And if you have specified blob endpoint for adls accounts, then it is corrected to its respective dfs endpoint.
- For user assigned MI, you can assign it to the Azure VM and in the config file, you can mention its `appid` in the azstorage section. Please try this with the latest release of blobfuse2.
- Can you give more details on what you mean by OIDC (federated credentials). For SPN, there is an option called `oauth-token-path` which points to the file where token is stored. It creates a workload identity credential which is generally used in AKS workspace. Can you check if this option works out for you.
- The azcli auth mode is merged in the main branch and the next release will have this feature, which will done around mid of May.

**Reply by rajsshah86** [reply]
Thanks @souravgupta-msft  for your reply

From MSI and SPN  login via openid connect  , I mean using the federated credential in MSI to login to azure and then using blobfuse.
<img width="1348" alt="image" src="https://github.com/Azure/azure-storage-fuse/assets/38250316/0fa4f8dd-2775-4512-bd64-02f8980986fa">

**Reply by rajsshah86** [reply]
To be clear on how we intend to use it
We have a  github workflow , where in we login via MSI (federated credentials) using [az login github action](https://github.com/marketplace/actions/azure-login). 

On further step we want to use blobfuse to mount the container on the githubrunner . We do not want to assign MI ( system / user defined) on the runner . Is there a way for it ?

For SPN As well same step as above .

**Reply by souravgupta-msft** [reply]
I will try this out. However, I think it is not supported in blobfuse currently.

**Reply by rajsshah86** [reply]
@souravgupta-msft : Thanks , I will wait for your response . 

Meanwhile , I have a feeling that above approach( Login using github action via MSI or SPN )  and then using blobfuse might work when I put mode as azcli  in the config file ??

**Reply by souravgupta-msft** [reply]
Yes, azcli auth mode should work for your scenario. It basically runs, `az account get-access-token` command. If this command runs successfully in the runner, then the blobfuse mount will also work.

**Reply by rajsshah86** [reply]
@souravgupta-msft : In that case I will have to wait for azcli mode to get officially released . However one more question then comes out of it , what will happen if i have a long running job ( more than an hour ) which is using the blobfuse2 mountpoint and the token gets expired.

**Reply by souravgupta-msft** [reply]
If token has expired, the token refresh logic will be executed where get-access-token command will be run to fetch the new token.
